### PR TITLE
Fix trait method visibility change leaking into other traits

### PR DESCRIPTION
**Pull request by ondrejmirtes**
… MemoizingSourceLocator

### CODE REVIEW

**Review by Ocramius — CHANGES_REQUESTED** [review]
Also here: `5.0.x` please :D

File: src/Reflection/ReflectionClass.php [hunk 1]
```diff
@@ -244,6 +244,7 @@ private function createMethodsFromTrait(ReflectionMethod $method): array
         $methodHash = $this->methodHash($method->getImplementingClass()->getName(), $method->getName());
 
         if (array_key_exists($methodHash, $traitModifiers)) {
+            $methodAst        = clone $methodAst;
```
Comment (Ocramius) [inline src/Reflection/ReflectionClass.php:247]: Oooooof, that's an ugly one :| 

I suppose AST should be marked `/** @psalm-immutable */` when possible, but too late for that...

File: test/unit/Reflection/ReflectionClassTest.php [hunk 2]
```diff
@@ -2482,4 +2483,61 @@ public function testGetAttributesByInstance(): void
...
+            interface Foo
+            {
+                public function doFoo(): void;
+            }
+
+            class Bar implements Foo
+            {
+                public function doFoo(): void
+                {
+                    echo "Doing foo the generic way";
```
Comment (Ocramius) [inline test/unit/Reflection/ReflectionClassTest.php]: Is it possible to reduce this code even further? I suppose method body isn't that relevant, right?

Comment (ondrejmirtes) [reply test/unit/Reflection/ReflectionClassTest.php]: Yeah, it's irrelevant, I just copied it from the PHPStan bug report :) Gonna amend that.

Comment (Ocramius) [reply test/unit/Reflection/ReflectionClassTest.php]: Figured: just want to have the test as-on-point-as-possible, before merging 👍

Comment (ondrejmirtes) [reply test/unit/Reflection/ReflectionClassTest.php]: Done, verified it still fails without the `clone`.

**Review by Ocramius — APPROVED** [review]
LGTM, thanks @ondrejmirtes!

**Pull request merged**